File: swift/Sources/PocketPrinter5890Kit/ReceiptRenderer.swift
import CoreGraphics
import CoreText
import Foundation

public enum ReceiptOrientation: String, CaseIterable, Identifiable, Sendable {
    case normal
    case rotated90

    public var id: String { rawValue }
    public var title: String { self == .normal ? "Normal" : "Rotation 90 deg" }
}

public enum DitherMode: String, CaseIterable, Identifiable, Sendable {
    case threshold
    case ordered
    case floydSteinberg

    public var id: String { rawValue }

    public var title: String {
        switch self {
        case .threshold: return "Seuil"
        case .ordered: return "Tramage ordonne"
        case .floydSteinberg: return "Floyd-Steinberg"
        }
    }
}

/// Rendu d'un ticket en bitmap monochrome.
///
/// Ecrit sur CoreGraphics et CoreText plutot qu'AppKit: le rendu fonctionne
/// ainsi a l'identique sur macOS, iOS et iPadOS. Une version anterieure
/// s'appuyait sur `NSImage`, ce qui restreignait la librairie a macOS et
/// exposait a un piege: sur ecran Retina, `NSImage.cgImage` renvoie une image
/// deux fois plus large en 64 bits par pixel, illisible telle quelle.
public struct ReceiptRenderer {
    public var width: Int
    public var threshold: UInt8
    public var ditherMode: DitherMode
    public var orientation: ReceiptOrientation

    public init(
        width: Int = MonochromeBitmap.nativeWidth,
        // Un ticket est fait de traits, pas de degrades. Floyd-Steinberg
        // disperse les pixels pour simuler du gris: applique a du texte, il
        // transforme chaque trait plein en semis de points, et le rendu sort
        // tres clair et tres fin. Le tramage ne vaut que pour une photo.
        //
        // Le seuil est relevé a 160: l'antialiasing du texte produit des bords
        // gris qu'un seuil trop bas efface, ce qui amaigrit le trait.
        threshold: UInt8 = 160,
        ditherMode: DitherMode = .threshold,
        orientation: ReceiptOrientation = .normal
    ) {
        self.width = width
        self.threshold = threshold
        self.ditherMode = ditherMode
        self.orientation = orientation
    }

    /// Compatibilite avec l'ancienne API booleenne.
    public init(width: Int, threshold: UInt8, dither: Bool, orientation: ReceiptOrientation = .normal) {
        self.init(
            width: width,
            threshold: threshold,
            ditherMode: dither ? .ordered : .threshold,
            orientation: orientation
        )
    }

    // MARK: - Rendu de ticket

    public func render(_ receipt: Receipt) throws -> MonochromeBitmap {
        let height = contentHeight(for: receipt)
        let gray = try drawToGrayscale(width: width, height: height) { context in
            drawReceipt(receipt, in: context, size: CGSize(width: width, height: height))
        }
        return try convert(gray: gray, width: width, height: height)
    }

    /// Image de previsualisation, utilisable directement par l'interface.
    public func previewImage(_ receipt: Receipt) throws -> CGImage {
        let height = contentHeight(for: receipt)
        return try drawToImage(width: width, height: height) { context in
            drawReceipt(receipt, in: context, size: CGSize(width: width, height: height))
        }
    }

    private func contentHeight(for receipt: Receipt) -> Int {
        90 + receipt.items.count * 22 + 80
    }

    // MARK: - Conversion d'image quelconque

    /// Convertit une image en bitmap a la largeur d'impression, proportions
    /// conservees.
    public func bitmap(from image: CGImage) throws -> MonochromeBitmap {
        guard image.width > 0, image.height > 0 else {
            throw RasterError.invalidDimensions
        }
        let scale = Double(width) / Double(image.width)
        let height = max(1, Int((Double(image.height) * scale).rounded()))

        let gray = try drawToGrayscale(width: width, height: height) { context in
            context.interpolationQuality = .high
            context.draw(image, in: CGRect(x: 0, y: 0, width: width, height: height))
        }
        return try convert(gray: gray, width: width, height: height)
    }

    // MARK: - Mire

    public func testPattern(height: Int = 240) throws -> MonochromeBitmap {
        let gray = try drawToGrayscale(width: width, height: height) { context in
            drawTestPattern(in: context, height: height)
        }
        return try convert(gray: gray, width: width, height: height)
    }

    // MARK: - Dessin

    private func drawReceipt(_ receipt: Receipt, in context: CGContext, size: CGSize) {
        let margin: CGFloat = 8
        let right = size.width - margin
        var y: CGFloat = 10

        drawText(receipt.merchantName.uppercased(), in: context,
                 at: .center(size.width), y: y, size: 22, bold: true)
        y += 30
        drawText(receipt.address, in: context, at: .center(size.width), y: y, size: 12)
        y += 18
        drawText(Self.dateFormatter.string(from: receipt.date), in: context,
                 at: .center(size.width), y: y, size: 12)
        y += 22

        strokeLine(in: context, from: margin, to: right, y: y)
        y += 10

        for item in receipt.items {
            drawText("\(item.quantity)x \(item.name)", in: context, at: .left(margin), y: y, size: 13)
            drawText(Self.money(item.total), in: context, at: .right(right), y: y, size: 13)
            y += 20
        }

        y += 4
        strokeLine(in: context, from: margin, to: right, y: y)
        y += 10

        drawText("TOTAL", in: context, at: .left(margin), y: y, size: 17, bold: true)
        drawText(Self.money(receipt.total), in: context, at: .right(right), y: y, size: 17, bold: true)
        y += 32

        drawText(receipt.footer, in: context, at: .center(size.width), y: y, size: 13)
    }

    private func drawTestPattern(in context: CGContext, height: Int) {
        let width = CGFloat(self.width)
        context.setStrokeColor(gray: 0, alpha: 1)
        context.setFillColor(gray: 0, alpha: 1)
        context.setLineWidth(1)

        // Cadre: verifie que toute la largeur est imprimee.
        context.stroke(CGRect(x: 0.5, y: 0.5, width: width - 1, height: CGFloat(height) - 1))

        // Graduation, reperes longs tous les 64 px.
        for x in stride(from: 0, to: self.width, by: 8) {
            let long = x % 64 == 0
            context.move(to: CGPoint(x: CGFloat(x), y: 0))
            context.addLine(to: CGPoint(x: CGFloat(x), y: long ? 20 : 10))
            context.strokePath()
        }

        // Damier de controle de densite.
        for row in 0..<6 {
            for column in 0..<12 where (row + column).isMultiple(of: 2) {
                context.fill(CGRect(x: 10 + column * 6, y: 34 + row * 6, width: 6, height: 6))
            }
        }

        // Aplat plein: revele une largeur mal cadree.
        context.fill(CGRect(x: 0, y: height - 30, width: Int(width), height: 12))

        drawText("\(self.width) px", in: context, at: .left(250), y: 44, size: 15, bold: true)
        drawText("GAUCHE", in: context, at: .left(4), y: CGFloat(height) - 52, size: 11)
        drawText("DROITE", in: context, at: .right(width - 4), y: CGFloat(height) - 52, size: 11)
    }

    // MARK: - Primitives de dessin

    private enum Anchor {
        case left(CGFloat)
        case right(CGFloat)
        case center(CGFloat)
    }

    private func strokeLine(in context: CGContext, from x1: CGFloat, to x2: CGFloat, y: CGFloat) {
        context.setStrokeColor(gray: 0, alpha: 1)
        context.setLineWidth(1)
        context.move(to: CGPoint(x: x1, y: y))
        context.addLine(to: CGPoint(x: x2, y: y))
        context.strokePath()
    }

    private func drawText(
        _ text: String,
        in context: CGContext,
        at anchor: Anchor,
        y: CGFloat,
        size: CGFloat,
        bold: Bool = false
    ) {
        // Police a chasse fixe: les colonnes d'un ticket doivent s'aligner.
        let font = CTFontCreateWithName(
            (bold ? "Menlo-Bold" : "Menlo-Regular") as CFString,
            size,
            nil
        )
        // Cles CoreText plutot qu'AppKit/UIKit: elles existent sur toutes
        // les plateformes.
        let attributes: [CFString: Any] = [
            kCTFontAttributeName: font,
            kCTForegroundColorAttributeName: CGColor(gray: 0, alpha: 1)
        ]
        let attributed = CFAttributedStringCreate(
            kCFAllocatorDefault,
            text as CFString,
            attributes as CFDictionary
        )
        guard let attributed else { return }
        let line = CTLineCreateWithAttributedString(attributed)
        let bounds = CTLineGetBoundsWithOptions(line, .useOpticalBounds)

        let x: CGFloat
        switch anchor {
        case .left(let value): x = value
        case .right(let value): x = value - bounds.width
        case .center(let total): x = (total - bounds.width) / 2
        }

        // Le contexte est retourne pour que la mise en page se lise de haut
        // en bas; CoreText dessinerait alors le texte en miroir. On annule
        // le retournement localement, le temps de tracer la ligne.
        context.saveGState()
        context.translateBy(x: 0, y: y + bounds.height)
        context.scaleBy(x: 1, y: -1)
        context.textPosition = CGPoint(x: x, y: 0)
        CTLineDraw(line, context)
        context.restoreGState()
    }

    // MARK: - Contexte graphique

    /// Dessine dans un contexte 8 bits par canal et renvoie les niveaux de gris.
    private func drawToGrayscale(
        width: Int,
        height: Int,
        _ draw: (CGContext) -> Void
    ) throws -> [UInt8] {
        guard width > 0, height > 0 else { throw RasterError.invalidDimensions }
        var pixels = [UInt8](repeating: 255, count: width * height)

        try pixels.withUnsafeMutableBytes { buffer in
            guard let context = CGContext(
                data: buffer.baseAddress,
                width: width,
                height: height,
                bitsPerComponent: 8,
                bytesPerRow: width,
                space: CGColorSpaceCreateDeviceGray(),
                bitmapInfo: CGImageAlphaInfo.none.rawValue
            ) else {
                throw RasterError.invalidDimensions
            }
            context.setFillColor(gray: 1, alpha: 1)
            context.fill(CGRect(x: 0, y: 0, width: width, height: height))

            // CoreGraphics place son origine en bas a gauche; l'imprimante
            // attend la premiere ligne en haut. On retourne le repere pour
            // que le code de mise en page se lise de haut en bas.
            context.translateBy(x: 0, y: CGFloat(height))
            context.scaleBy(x: 1, y: -1)
            draw(context)
        }
        return pixels
    }

    private func drawToImage(
        width: Int,
        height: Int,
        _ draw: (CGContext) -> Void
    ) throws -> CGImage {
        guard let context = CGContext(
            data: nil,
            width: width,
            height: height,
            bitsPerComponent: 8,
            bytesPerRow: 0,
            space: CGColorSpaceCreateDeviceGray(),
            bitmapInfo: CGImageAlphaInfo.none.rawValue
        ) else {
            throw RasterError.invalidDimensions
        }
        context.setFillColor(gray: 1, alpha: 1)
        context.fill(CGRect(x: 0, y: 0, width: width, height: height))
        context.translateBy(x: 0, y: CGFloat(height))
        context.scaleBy(x: 1, y: -1)
        draw(context)

        guard let image = context.makeImage() else {
            throw RasterError.invalidDimensions
        }
        return image
    }

    private func convert(gray: [UInt8], width: Int, height: Int) throws -> MonochromeBitmap {
        if orientation == .rotated90 {
            var rotated = [UInt8](repeating: 255, count: width * height)
            for y in 0..<height {
                for x in 0..<width {
                    rotated[x * height + (height - 1 - y)] = gray[y * width + x]
                }
            }
            return try binarise(gray: rotated, width: height, height: width)
        }
        return try binarise(gray: gray, width: width, height: height)
    }

    private func binarise(gray: [UInt8], width: Int, height: Int) throws -> MonochromeBitmap {
        switch ditherMode {
        case .threshold:
            return try MonochromeConverter.threshold(
                grayscale: gray, width: width, height: height, threshold: threshold
            )
        case .ordered:
            return try MonochromeConverter.orderedDither(grayscale: gray, width: width, height: height)
        case .floydSteinberg:
            return try MonochromeConverter.floydSteinberg(grayscale: gray, width: width, height: height)
        }
    }

    private static let dateFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.dateStyle = .short
        formatter.timeStyle = .short
        return formatter
    }()

    private static func money(_ value: Decimal) -> String {
        let formatter = NumberFormatter()
        formatter.numberStyle = .currency
        formatter.currencyCode = "EUR"
        formatter.maximumFractionDigits = 2
        return formatter.string(from: value as NSDecimalNumber) ?? "\(value)"
    }
}

File: android/render/src/main/kotlin/fr/pocketprinter5890/render/ReceiptRenderer.kt
package fr.pocketprinter5890.render

import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Typeface
import fr.pocketprinter5890.kit.MonochromeBitmap
import fr.pocketprinter5890.kit.MonochromeConverter
import fr.pocketprinter5890.kit.Receipt
import fr.pocketprinter5890.kit.RasterException
import fr.pocketprinter5890.kit.formatCents
import java.text.SimpleDateFormat
import java.util.Locale

/** Orientation du rendu. */
enum class ReceiptOrientation(val title: String) {
    NORMAL("Normal"),
    ROTATED_90("Rotation 90 deg")
}

/** Methode de conversion des niveaux de gris en noir et blanc. */
enum class DitherMode(val title: String) {
    /** Seuil dur. Le seul valable pour un code a scanner. */
    THRESHOLD("Seuil"),
    ORDERED("Tramage ordonne"),
    FLOYD_STEINBERG("Floyd-Steinberg")
}

/**
 * Rendu d'un ticket en bitmap monochrome.
 *
 * Equivalent Android de `ReceiptRenderer` cote Swift: meme mise en page,
 * memes proportions, dessine ici avec `android.graphics` au lieu de
 * CoreGraphics.
 *
 * C'est le chemin qu'emprunte l'application officielle pour **tout** son
 * texte, ce qui lui permet le gras et les accents que le firmware ne sait pas
 * rendre lui-meme. En contrepartie le resultat est plus doux que la police
 * interne, et un ticket pese des milliers d'octets au lieu de quelques
 * dizaines.
 */
class ReceiptRenderer(
    val width: Int = MonochromeBitmap.nativeWidth,
    /**
     * Seuil de noircissement, 0 a 255.
     *
     * Un pixel plus sombre que cette valeur est imprime en noir. Monter le
     * seuil epaissit le rendu: utile quand le texte antialiase sort trop
     * maigre.
     */
    val threshold: Int = 160,
    /**
     * Defaut: seuil dur.
     *
     * Un ticket est fait de traits, pas de degrades. Floyd-Steinberg disperse
     * les pixels pour simuler du gris: applique a du texte, il transforme
     * chaque trait plein en semis de points et le rendu sort tres clair et
     * tres fin. Le tramage n'a d'interet que pour une photo.
     */
    val ditherMode: DitherMode = DitherMode.THRESHOLD,
    val orientation: ReceiptOrientation = ReceiptOrientation.NORMAL
) {

    // MARK: - Rendu de ticket

    /** Rend le ticket en bitmap pret a etre envoye en raster. */
    @Throws(RasterException::class)
    fun render(receipt: Receipt): MonochromeBitmap {
        val image = renderToBitmap(receipt)
        return try {
            convert(image)
        } finally {
            image.recycle()
        }
    }

    /**
     * Image de previsualisation, affichable directement par l'interface.
     *
     * Elle passe par la meme binarisation que l'impression: un apercu en
     * niveaux de gris montrerait un rendu plus dense que le papier, et le
     * reglage du seuil n'y serait pas visible.
     *
     * L'appelant est responsable de son cycle de vie.
     */
    fun previewImage(receipt: Receipt): Bitmap {
        val drawn = renderToBitmap(receipt)
        return try {
            toDisplayBitmap(convert(drawn))
        } finally {
            drawn.recycle()
        }
    }

    /** Reconstruit une image affichable a partir du bitmap monochrome. */
    private fun toDisplayBitmap(source: MonochromeBitmap): Bitmap {
        val pixels = IntArray(source.width * source.height)
        for (y in 0 until source.height) {
            for (x in 0 until source.width) {
                val byte = source.bytes[y * source.widthBytes + x / 8].toInt() and 0xff
                val on = (byte shr (7 - x % 8)) and 1 == 1
                pixels[y * source.width + x] = if (on) Color.BLACK else Color.WHITE
            }
        }
        return Bitmap.createBitmap(
            pixels, source.width, source.height, Bitmap.Config.ARGB_8888
        )
    }

    private fun renderToBitmap(receipt: Receipt): Bitmap {
        val height = contentHeight(receipt)
        val image = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(image)
        canvas.drawColor(Color.WHITE)
        drawReceipt(receipt, canvas, height)
        return image
    }

    private fun contentHeight(receipt: Receipt): Int =
        90 + receipt.items.size * 22 + 80

    // MARK: - Conversion d'image quelconque

    /**
     * Convertit une image a la largeur d'impression, proportions conservees.
     *
     * Le filtrage est actif ici: pour une photo il ameliore le rendu. A
     * proscrire en revanche pour un code a scanner, dont les bords doivent
     * rester francs.
     */
    @Throws(RasterException::class)
    fun bitmap(source: Bitmap): MonochromeBitmap {
        if (source.width <= 0 || source.height <= 0) {
            throw RasterException.InvalidDimensions
        }
        val scale = width.toDouble() / source.width
        val height = maxOf(1, Math.round(source.height * scale).toInt())
        val scaled = Bitmap.createScaledBitmap(source, width, height, true)
        return try {
            convert(scaled)
        } finally {
            if (scaled !== source) scaled.recycle()
        }
    }

    // MARK: - Mire

    /** Mire de controle: cadre, graduation, damier de densite, aplat plein. */
    @Throws(RasterException::class)
    fun testPattern(height: Int = 240): MonochromeBitmap {
        val image = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(image)
        canvas.drawColor(Color.WHITE)
        drawTestPattern(canvas, height)
        return try {
            convert(image)
        } finally {
            image.recycle()
        }
    }

    // MARK: - Dessin

    private fun drawReceipt(receipt: Receipt, canvas: Canvas, height: Int) {
        val margin = 8f
        val right = width - margin
        var y = 10f

        drawText(
            receipt.merchantName.uppercase(Locale.FRANCE), canvas,
            Anchor.Center, y, 22f, bold = true
        )
        y += 30f
        if (receipt.address.isNotEmpty()) {
            drawText(receipt.address, canvas, Anchor.Center, y, 12f)
        }
        y += 18f
        drawText(dateFormatter.format(receipt.date), canvas, Anchor.Center, y, 12f)
        y += 22f

        strokeLine(canvas, margin, right, y)
        y += 10f

        for (item in receipt.items) {
            drawText("${item.quantity}x ${item.name}", canvas, Anchor.Left(margin), y, 13f)
            drawText(formatCents(item.totalCents), canvas, Anchor.Right(right), y, 13f)
            y += 20f
        }

        y += 4f
        strokeLine(canvas, margin, right, y)
        y += 10f

        // Le gras est ici reellement dessine: c'est tout l'interet du chemin
        // image, le firmware ne sachant pas appliquer `ESC E`.
        drawText("TOTAL", canvas, Anchor.Left(margin), y, 17f, bold = true)
        drawText(
            formatCents(receipt.totalCents), canvas,
            Anchor.Right(right), y, 17f, bold = true
        )
        y += 32f

        if (receipt.footer.isNotEmpty()) {
            drawText(receipt.footer, canvas, Anchor.Center, y, 13f)
        }
    }

    private fun drawTestPattern(canvas: Canvas, height: Int) {
        val paint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
            color = Color.BLACK
            strokeWidth = 1f
        }

        // Cadre: verifie que toute la largeur est imprimee.
        paint.style = Paint.Style.STROKE
        canvas.drawRect(0.5f, 0.5f, width - 0.5f, height - 0.5f, paint)

        // Graduation, reperes longs tous les 64 px.
        var x = 0
        while (x < width) {
            val long = x % 64 == 0
            canvas.drawLine(x.toFloat(), 0f, x.toFloat(), if (long) 20f else 10f, paint)
            x += 8
        }

        // Damier de controle de densite.
        paint.style = Paint.Style.FILL
        for (row in 0 until 6) {
            for (column in 0 until 12) {
                if ((row + column) % 2 != 0) continue
                val left = (10 + column * 6).toFloat()
                val top = (34 + row * 6).toFloat()
                canvas.drawRect(left, top, left + 6f, top + 6f, paint)
            }
        }

        // Aplat plein: revele une largeur mal cadree.
        canvas.drawRect(0f, (height - 30).toFloat(), width.toFloat(), (height - 18).toFloat(), paint)

        drawText("$width px", canvas, Anchor.Left(250f), 44f, 15f, bold = true)
        drawText("GAUCHE", canvas, Anchor.Left(4f), height - 52f, 11f)
        drawText("DROITE", canvas, Anchor.Right(width - 4f), height - 52f, 11f)
    }

    // MARK: - Primitives de dessin

    private sealed class Anchor {
        data class Left(val x: Float) : Anchor()
        data class Right(val x: Float) : Anchor()
        object Center : Anchor()
    }

    private fun strokeLine(canvas: Canvas, from: Float, to: Float, y: Float) {
        val paint = Paint().apply {
            color = Color.BLACK
            strokeWidth = 1f
        }
        canvas.drawLine(from, y, to, y, paint)
    }

    private fun drawText(
        text: String,
        canvas: Canvas,
        anchor: Anchor,
        y: Float,
        size: Float,
        bold: Boolean = false
    ) {
        // Police a chasse fixe: les colonnes d'un ticket doivent s'aligner.
        //
        // L'antialiasing est conserve car il rend les formes des lettres, mais
        // il produit des bords gris que le seuil supprime, ce qui amaigrit le
        // trait. `isFakeBoldText` compense en epaississant le glyphe lui-meme:
        // le noir reellement pose sur le papier augmente, la ou `Typeface.BOLD`
        // seul reste trop fin une fois binarise.
        val paint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
            color = Color.BLACK
            textSize = size
            typeface = Typeface.create(
                Typeface.MONOSPACE,
                if (bold) Typeface.BOLD else Typeface.NORMAL
            )
            isFakeBoldText = bold
        }
        val measured = paint.measureText(text)
        val x = when (anchor) {
            is Anchor.Left -> anchor.x
            is Anchor.Right -> anchor.x - measured
            Anchor.Center -> (width - measured) / 2f
        }
        // `y` designe le haut du texte, comme cote Swift: on decale de
        // l'ascendante pour que les deux rendus se superposent.
        canvas.drawText(text, x, y - paint.fontMetrics.ascent, paint)
    }

    // MARK: - Conversion

    @Throws(RasterException::class)
    private fun convert(image: Bitmap): MonochromeBitmap {
        val source = if (orientation == ReceiptOrientation.ROTATED_90) {
            rotate90(image)
        } else {
            image
        }
        try {
            val imageWidth = source.width
            val imageHeight = source.height
            val pixels = IntArray(imageWidth * imageHeight)
            source.getPixels(pixels, 0, imageWidth, 0, 0, imageWidth, imageHeight)

            // Luminance perceptuelle: un gris moyen naif noircit les rouges.
            val gray = IntArray(pixels.size) { index ->
                val pixel = pixels[index]
                val red = (pixel shr 16) and 0xff
                val green = (pixel shr 8) and 0xff
                val blue = pixel and 0xff
                (red * 299 + green * 587 + blue * 114) / 1000
            }

            return when (ditherMode) {
                DitherMode.THRESHOLD ->
                    MonochromeConverter.threshold(gray, imageWidth, imageHeight, threshold)
                DitherMode.ORDERED ->
                    MonochromeConverter.orderedDither(gray, imageWidth, imageHeight)
                DitherMode.FLOYD_STEINBERG ->
                    MonochromeConverter.floydSteinberg(gray, imageWidth, imageHeight)
            }
        } finally {
            if (source !== image) source.recycle()
        }
    }

    /**
     * Rotation d'un quart de tour.
     *
     * La largeur devient l'ancienne hauteur, qui n'a aucune raison d'etre un
     * multiple de 8. Comme un bitmap monochrome exige des lignes entieres en
     * octets, on complete a droite en blanc jusqu'au multiple suivant plutot
     * que de laisser echouer la conversion.
     */
    private fun rotate90(image: Bitmap): Bitmap {
        val matrix = android.graphics.Matrix().apply { postRotate(90f) }
        val rotated = Bitmap.createBitmap(image, 0, 0, image.width, image.height, matrix, true)
        if (rotated.width % 8 == 0) return rotated

        val padded = ((rotated.width + 7) / 8) * 8
        val output = Bitmap.createBitmap(padded, rotated.height, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(output)
        canvas.drawColor(Color.WHITE)
        canvas.drawBitmap(rotated, 0f, 0f, null)
        rotated.recycle()
        return output
    }

    private companion object {
        val dateFormatter = SimpleDateFormat("dd/MM/yyyy HH:mm", Locale.FRANCE)
    }
}
